### PR TITLE
[DAGCombiner] Add basic support for `trunc nsw/nuw`

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch adds basic support for `trunc nsw/nuw` in SDAG. It will allow DAGCombiner to further eliminate in-reg `zext/sext` instructions.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -14083,24 +14084,23 @@ SDValue DAGCombiner::visitZERO_EXTEND(SDNode *N) {
...
       unsigned MidBits = MinVT.getScalarSizeInBits();
       unsigned DestBits = VT.getScalarSizeInBits();
-      unsigned NumSignBits = DAG.ComputeNumSignBits(Op);
 
-      if (OpBits == DestBits) {
-        // Op is i32, Mid is i8, and Dest is i32.  If Op has more than 24 sign
-        // bits, it is already ready.
-        if (NumSignBits > DestBits - MidBits)
+      if (N0->getFlags().hasNoSignedWrap() ||
+          DAG.ComputeNumSignBits(Op) > OpBits - MidBits) {
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:14089]: Should probably just set `nsw` if `visitTrunc` if `computeNumSignBits >= OldSize - NewSize`
Likewise `nuw` if high bits are zero.

**Comment by dtcxzyw** [comment]
Miscompilation:
```
define i32 @test(i1 zeroext %x, i1 zeroext %y) {
entry:
  %sel = select i1 %y, i64 4, i64 0
  %conv0 = sext i1 %x to i64
  %xor = xor i64 %sel, %conv0
  %conv1 = trunc nsw i64 %xor to i32
  %div = sdiv i32 %conv1, -10765
  ret i32 %div
}
```
Before:
```
Optimized legalized selection DAG: %bb.0 'test:entry'
SelectionDAG has 27 nodes:
  t0: ch,glue = EntryToken
            t7: i32,ch = CopyFromReg t0, Register:i32 %1
          t71: i32 = AssertZext t7, ValueType:ch:i1
        t57: i32 = shl t71, Constant:i8<2>
            t2: i32,ch = CopyFromReg t0, Register:i32 %0
          t73: i32 = AssertZext t2, ValueType:ch:i1
        t78: i32 = sub Constant:i32<0>, t73
      t48: i32 = xor t57, t78
    t62: i64 = sign_extend t48
  t42: i64 = mul t62, Constant:i64<-1634202141>
        t64: i64 = sra t42, Constant:i8<44>
      t65: i32 = truncate t64
        t68: i64 = srl t42, Constant:i8<63>
      t69: i32 = truncate t68
    t37: i32 = add t65, t69
  t25: ch,glue = CopyToReg t0, Register:i32 $eax, t37
  t26: ch = X86ISD::RET_GLUE t25, TargetConstant:i32<0>, Register:i32 $eax, t25:1
```
After:
```
Optimized legalized selection DAG: %bb.0 'test:entry'
SelectionDAG has 27 nodes:
  t0: ch,glue = EntryToken
            t7: i32,ch = CopyFromReg t0, Register:i32 %1
          t66: i32 = AssertZext t7, ValueType:ch:i1
        t57: i32 = shl t66, Constant:i8<2>
            t2: i32,ch = CopyFromReg t0, Register:i32 %0
          t68: i32 = AssertZext t2, ValueType:ch:i1
        t73: i32 = sub Constant:i32<0>, t68
      t48: i32 = xor t57, t73
    t49: i64 = any_extend t48
  t42: i64 = mul t49, Constant:i64<-1634202141>
        t61: i64 = sra t42, Constant:i8<44>
      t62: i32 = truncate t61
        t64: i64 = srl t42, Constant:i8<63>
      t65: i32 = truncate t64
    t37: i32 = add t62, t65
  t25: ch,glue = CopyToReg t0, Register:i32 $eax, t37
  t26: ch = X86ISD::RET_GLUE t25, TargetConstant:i32<0>, Register:i32 $eax, t25:1
```

**Comment by dtcxzyw** [comment]
> Miscompilation:
> 
> ```
> define i32 @test(i1 zeroext %x, i1 zeroext %y) {
> entry:
>   %sel = select i1 %y, i64 4, i64 0
>   %conv0 = sext i1 %x to i64
>   %xor = xor i64 %sel, %conv0
>   %conv1 = trunc nsw i64 %xor to i32
>   %div = sdiv i32 %conv1, -10765
>   ret i32 %div
> }
> ```
> 
> Before:
> 
> ```
> Optimized legalized selection DAG: %bb.0 'test:entry'
> SelectionDAG has 27 nodes:
>   t0: ch,glue = EntryToken
>             t7: i32,ch = CopyFromReg t0, Register:i32 %1
>           t71: i32 = AssertZext t7, ValueType:ch:i1
>         t57: i32 = shl t71, Constant:i8<2>
>             t2: i32,ch = CopyFromReg t0, Register:i32 %0
>           t73: i32 = AssertZext t2, ValueType:ch:i1
>         t78: i32 = sub Constant:i32<0>, t73
>       t48: i32 = xor t57, t78
>     t62: i64 = sign_extend t48
>   t42: i64 = mul t62, Constant:i64<-1634202141>
>         t64: i64 = sra t42, Constant:i8<44>
>       t65: i32 = truncate t64
>         t68: i64 = srl t42, Constant:i8<63>
>       t69: i32 = truncate t68
>     t37: i32 = add t65, t69
>   t25: ch,glue = CopyToReg t0, Register:i32 $eax, t37
>   t26: ch = X86ISD::RET_GLUE t25, TargetConstant:i32<0>, Register:i32 $eax, t25:1
> ```
> 
> After:
> 
> ```
> Optimized legalized selection DAG: %bb.0 'test:entry'
> SelectionDAG has 27 nodes:
>   t0: ch,glue = EntryToken
>             t7: i32,ch = CopyFromReg t0, Register:i32 %1
>           t66: i32 = AssertZext t7, ValueType:ch:i1
>         t57: i32 = shl t66, Constant:i8<2>
>             t2: i32,ch = CopyFromReg t0, Register:i32 %0
>           t68: i32 = AssertZext t2, ValueType:ch:i1
>         t73: i32 = sub Constant:i32<0>, t68
>       t48: i32 = xor t57, t73
>     t49: i64 = any_extend t48
>   t42: i64 = mul t49, Constant:i64<-1634202141>
>         t61: i64 = sra t42, Constant:i8<44>
>       t62: i32 = truncate t61
>         t64: i64 = srl t42, Constant:i8<63>
>       t65: i32 = truncate t64
>     t37: i32 = add t62, t65
>   t25: ch,glue = CopyToReg t0, Register:i32 $eax, t37
>   t26: ch = X86ISD::RET_GLUE t25, TargetConstant:i32<0>, Register:i32 $eax, t25:1
> ```

These flags should be cleared in `TargetLowering::SimplifyDemandedBits`.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 2]
```diff
@@ -2572,14 +2572,23 @@ bool TargetLowering::SimplifyDemandedBits(
...
     APInt TruncMask = DemandedBits.zext(OperandBitWidth);
     if (SimplifyDemandedBits(Src, TruncMask, DemandedElts, Known, TLO,
-                             Depth + 1))
+                             Depth + 1)) {
+      if (Flags.hasNoSignedWrap() || Flags.hasNoUnsignedWrap()) {
+        // Disable the nsw and nuw flags. We can no longer guarantee that we
+        // won't wrap after simplification.
+        Flags.setNoSignedWrap(false);
+        Flags.setNoUnsignedWrap(false);
+        Op->setFlags(Flags);
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:2588]: The API for SDNodeFlags is much worse than FastMathFlags. Can we get a wrapper with bit operations so you can just do a ~WrapFlags?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 3]
```diff
@@ -13807,23 +13809,25 @@ SDValue DAGCombiner::visitSIGN_EXTEND(SDNode *N) {
...
+        DAG.ComputeNumSignBits(Op) > OpBits - MidBits) {
+      if (OpBits == DestBits) {
+        // Op is i32, Mid is i8, and Dest is i32.  If Op has more than 24 sign
+        // bits, it is already ready.
         return Op;
-    } else if (OpBits < DestBits) {
-      // Op is i32, Mid is i8, and Dest is i64.  If Op has more than 24 sign
-      // bits, just sext from i32.
-      if (NumSignBits > OpBits-MidBits)
+      } else if (OpBits < DestBits) {
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:13819]: Drop else after return?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 4]
```diff
@@ -13807,23 +13809,27 @@ SDValue DAGCombiner::visitSIGN_EXTEND(SDNode *N) {
...
     unsigned MidBits  = N0.getScalarValueSizeInBits();
     unsigned DestBits = VT.getScalarSizeInBits();
-    unsigned NumSignBits = DAG.ComputeNumSignBits(Op);
 
-    if (OpBits == DestBits) {
-      // Op is i32, Mid is i8, and Dest is i32.  If Op has more than 24 sign
-      // bits, it is already ready.
-      if (NumSignBits > DestBits-MidBits)
+    if (N0->getFlags().hasNoSignedWrap() ||
+        DAG.ComputeNumSignBits(Op) > OpBits - MidBits) {
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:13814]: Still think the way to do this is just set `nsw`/`nuw` in `visitTRUNC` and drop the `compute...`

Comment (dtcxzyw) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:13814]: Unlike in InstCombine, we don't infer poison generating flags in SDAG. Without recursive analysis we may miss some optimization opportunities.

Comment (goldsteinn) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:13814]: I guess you're right. Is there a good reason for that?

**Comment by dtcxzyw** [comment]
Ping.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 5]
```diff
@@ -2552,14 +2552,19 @@ bool TargetLowering::SimplifyDemandedBits(
   }
   case ISD::TRUNCATE: {
     SDValue Src = Op.getOperand(0);
+    SDNodeFlags Flags = Op->getFlags();
```
Comment (nikic) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:2555]: Unused Flags variable?

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**